sympy/polys/matrices/sdm.py

```diff
@@ -28,6 +28,17 @@ def __init__(self, elemsdict, shape, domain):
         if not all(0 <= c < n for row in self.values() for c in row):
             raise DDMBadInputError("Column out of range")
 
+    def extract_slice(self, slice1, slice2):
+        r = slice1.stop-slice1.start
+        c = slice2.stop-slice2.start
+
+        def extract_map(value):
+            return (value[0], {i: value[1][j] for i, j in enumerate(slice2)})
+
+        val = {i: self[j] for i, j in enumerate(slice1)}
+        sdm = dict(map(extract_map, val.items()))
```

Comment:
For something like `M[::-1, :]` the slices are the same size as the dimensions of the matrix. That means that this will iterate over all rows and columns even if there are no elements in that row and column.

You need to understand that an SDM can be an enormous matrix but have very few elements e.g.:
```python
In [4]: dM = DomainMatrix({0:{0:ZZ(1)}}, (1000000, 1000000), ZZ)

In [5]: dM.shape
Out[5]: (1000000, 1000000)

In [6]: dM.rep
Out[6]: {0: {0: mpz(1)}}

In [7]: type(_)
Out[7]: sympy.polys.matrices.sdm.SDM
```
Operations with `SDM` need to be carefully optimised so that they only ever loop over the nonzero elements of the matrix. The matrix `dM` above is a 1000000x1000000 matrix but the sparse algorithms that work with it can return in microseconds:
```python
In [8]: %time ok = dM**2
CPU times: user 94 µs, sys: 1e+03 ns, total: 95 µs
Wall time: 100 µs

In [9]: %time ok = dM+dM
CPU times: user 76 µs, sys: 1 µs, total: 77 µs
Wall time: 83 µs

In [11]: %time ok = dM.to_field().rref()
CPU times: user 199 µs, sys: 11 µs, total: 210 µs
Wall time: 226 µs
```
These operations would be much slower if they involved *any* loop over either the million rows or the million columns. The implementation above loops over both in a *nested* loop which means `10^12` calls to `value[1][j]`.

To implement `dM[::-1, :]` efficiently it must not iterate over the slices. Instead the slice can be converted to a range `r` and then you can do something like:
```
rows = {k: v for k, v in self.items() if k in r}
```
That loops over the nonzero rows only so for `dM` above that means a single iteration.

There might be some way to make that more efficient using `map` and the primitives in the operator module but note that:
1. Getting the big-O behaviour correct is the first consideration: it should be `O(k)` where `k` is the number of nonzero elements rather than `O(N^2)` where N is the number of rows/columns.
2. Using `map` with a Python function is not generally more efficient than e.g. a dict comprehension. The speedup from using `map` is when it is used with callables and iterators that are all implemented in C.

Whatever you do you need to run timings to see how fast it is.

Reply:
My suggestion for `extract_slice` would be:
```python
    def extract_slice(self, slice1, slice2):
        m, n = self.shape
        ri = range(m)[slice1]
        ci = range(n)[slice2]

        sdm = {}
        for i, row in self.items():
            if i in ri:
                row = {ci.index(j): e for j, e in row.items() if j in ci}
                sdm[ri.index(i)] = row

        return self.new(sdm, (len(ri), len(ci)), self.domain)
```
Note that `slice1` and `slice2` are slices and not `range`.

Reply:
> You need to understand that an SDM can be an enormous matrix but have very few elements e.g.:
> 
I didn't realize the scale could be so much. Will keep this in mind for future implementation.

> 1. Getting the big-O behaviour correct is the first consideration: it should be `O(k)` where `k` is the number of nonzero elements rather than `O(N^2)` where N is the number of rows/columns.
>

Sure, I have used the implementation you gave and it seems to be too fast:
```
In [1]: from sympy.polys.matrices import DomainMatrix

In [2]: dM = DomainMatrix({0:{0:ZZ(1)}}, (1000000, 1000000), ZZ)

In [3]: %time ok = dM[::-1, :]
Wall time: 0 ns
```
Thanks for all the help. I was busy this week with the academics and some presentations. Will be free from now, and more active from now on.

Reply:
> My suggestion for `extract_slice` would be:
> 
> ```python
>     def extract_slice(self, slice1, slice2):
>         m, n = self.shape
>         ri = range(m)[slice1]
>         ci = range(n)[slice2]
> 
>         sdm = {}
>         for i, row in self.items():
>             if i in ri:
>                 row = {ci.index(j): e for j, e in row.items() if j in ci}
>                 sdm[ri.index(i)] = row
> 
>         return self.new(sdm, (len(ri), len(ci)), self.domain)
> ```
> 
> Note that `slice1` and `slice2` are slices and not `range`.

Sure, this just works the bestm have implemented this for `SDM`.